util/election2/election.go

```diff
@@ -0,0 +1,84 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package election2 provides master election tools, and interfaces for
+// plugging in a custom underlying mechanism.
+//
+// There are two important abstractions in this package: instance and resource.
+// - An instance is a single client of the library. An instance is represented
+//   by an Election object.
+// - A resource is something guarded by master election (e.g. a piece of data,
+//   or operation). Each resource has at most one (most of the time; see note
+//   below) master instance which is said to own this resource. A single
+//   instance may own multiple resources.
+//
+// Note: Sometimes there can be more than 1 instance "believing" to own a
+// resource. The reason is that the client code operates outside of the
+// election mechanism (e.g. a distributed consensus group), so mastership
+// updates can race with the operation.
+//
+// TODO(pavelkalinnikov): Merge this package with util/election.
+package election2
+
+import "context"
+
+// Election controls an instance's participation in master election process.
+// Note: Implementations are not intended to be thread-safe.
+type Election interface {
+	// Await blocks until the instance captures mastership. Returns immediately
+	// if it is already the master. Returns an error if capturing fails, or the
+	// passed in context is canceled before mastership is captured. If an error
+	// is returned, the instance might still have become the master. Idempotent,
+	// might be useful to retry in case of an error.
+	Await(ctx context.Context) error
+
+	// Observe returns a "mastership context" which remains active until the
+	// instance stops being the master, or the passed in context is canceled. If
+	// the instance is not the master during this call, returns an already
+	// canceled context. In particular, this will happen if Observe is called
+	// without a preceding Await.
+	//
+	// The resources used for maintaining the mastership context are released
+	// when the latter gets canceled. This happens when the instance loses
+	// mastership, calls Resign, an error occurs in mastership monitoring, or the
+	// context passed in to Observe is explicitly canceled.
+	Observe(ctx context.Context) (context.Context, error)
+
+	// Resign releases mastership for this instance. The instance can be elected
+	// again using Await. Idempotent, might be useful to retry if fails.
+	//
+	// Note: Resign does not guarantee immediate cancelation of the context
+	// returned from Observe. However, the latter will happen *eventually* if
+	// resigning is successful. The caller can force mastership context
+	// cancelation by explicitly canceling the context passed in to Observe.
```

Comment:
There is a trade-off here.

We could enforce that `Election` cancels mastership context returned from `WithMastership` right away when `Resign` is called, but that complicates the implementation:
- `Observe/WithMastership` must store `cancel` function in the `Election` object to be able to cancel it from `Resign`
- It implies various problems like:
  - do we have to guard it with a mutex?
  - what if multiple `WithMastership` calls happen? do we store another `cancel` after canceling the previous one, or do we return an error?
  - what if `Resign` is called with no preceding `WithMastership`?
  - etc

Basically we have to handle a lot of additional corner cases (in all potential `Election` interface implementations), only for the sake of synchronous context cancelation. **But** what we get in return is a rather weak benefit, and here is why. I think it would be a bad practice for the caller to rely on synchronous mastership context cancelation as a normal way of finishing its work upon resigning, mainly because it's hard to finish any work correctly under a canceled context. A better practice is as follows:
- Finish work while mastership context is still alive. While we are the master we are most likely to be able to do it correctly/gracefully.
- Now we don't care about the mastership context anymore, thus we don't care that it will be canceled asynchronously.
- Call `Resign`. It will result in resignation, plus an **asynchronous** mastership context cancelation (which is there only for releasing the resources associated with that context).